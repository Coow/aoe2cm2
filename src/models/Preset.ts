import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);
    public readonly name: string;
    public presetId?: string
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = [], presetId?: string) {
        this.name = name;
        this.presetId = presetId;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[], presetId?: string } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        Assert.isOptionalString(preset.presetId);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns), preset.presetId);
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;