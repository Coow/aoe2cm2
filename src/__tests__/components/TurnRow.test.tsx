import {shallow} from "enzyme";
import TurnRow from "../../components/draft/TurnRow";
import Preset from "../../models/Preset";


it('TurnRow renders correctly', () => {
    const component = shallow(<TurnRow turns={Preset.SAMPLE.turns}/>);
    expect(component).toMatchSnapshot();
});
