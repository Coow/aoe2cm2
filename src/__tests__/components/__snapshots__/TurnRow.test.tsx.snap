// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurnRow renders correctly 1`] = `
<TurnRow
  t={[Function]}
  turns={
    Array [
      Turn {
        "action": "BAN",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "HOST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "BAN",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "GUEST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "BAN",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "BAN",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "REVEAL_ALL",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "NONE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "NONE",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "HOST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "GUEST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "EXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "EXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "PICK",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "REVEAL_ALL",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "NONE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "NONE",
      },
      Turn {
        "action": "SNIPE",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Turn {
        "action": "SNIPE",
        "categories": Array [
          "default",
        ],
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": true,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Turn {
        "action": "REVEAL_ALL",
        "categories": Array [
          "default",
        ],
        "exclusivity": "GLOBAL",
        "executingPlayer": "NONE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "NONE",
      },
    ]
  }
/>
`;
